windows-gaming-driver/src/driver/clipboard.rs

```diff
@@ -0,0 +1,213 @@
+                    Async::NotReady
+                }),
+            },
+            Async::NotReady => Ok(Async::NotReady),
+        }
+    }
+}
+
+fn get_atom(connection: &Connection, name: &str) -> Result<Atom, GenericError> {
+    xcb::intern_atom(connection, false, name).get_reply().map(|reply| reply.atom())
+}
+
+struct Atoms {
+    clipboard: Atom,
+    targets: Atom,
+    utf8_string: Atom,
+    property: Atom,
+}
+
+pub struct X11Clipboard {
+    connection: Connection,
+    window: u32,
+    atoms: Atoms,
+}
+
+pub struct ClipboardContext<'a> {
+    pub listener: Box<Future<Item=(), Error=::std::io::Error> + 'a>,
+    pub lost_recv: UnboundedReceiver<()>,
+    pub query_recv: UnboundedReceiver<ClipboardRequestEvent>,
+    pub data_recv: UnboundedReceiver<Vec<u8>>,
+    pub resp_send: UnboundedSender<ClipboardRequestResponse>,
+}
+
+impl X11Clipboard {
+    pub fn open() -> Result<X11Clipboard, GenericError> {
+        let (connection, screen) = Connection::connect(None).unwrap();
+
+        let window = connection.generate_id();
+
+        // borrowing a little from the x11-clipboard crate here
+        {
+            let screen = connection.get_setup().roots().nth(screen as usize)
+                .expect("Invalid X11 screen!");
+
+            xcb::create_window(
+                &connection,
+                xcb::COPY_FROM_PARENT as u8,
+                window, screen.root(),
+                0, 0, 1, 1,
+                0,
+                xcb::WINDOW_CLASS_INPUT_OUTPUT as u16,
+                screen.root_visual(),
+                &[(
+                    xcb::CW_EVENT_MASK,
+                    xcb::EVENT_MASK_STRUCTURE_NOTIFY | xcb::EVENT_MASK_PROPERTY_CHANGE
+                )]
+            );
+            connection.flush();
+        }
+
+        let clipboard = get_atom(&connection, "CLIPBOARD")?;
+        let property = get_atom(&connection, "THIS_CLIPBOARD_OUT")?;
+        let targets = get_atom(&connection, "TARGETS")?;
+        let utf8_string = get_atom(&connection, "UTF8_STRING")?;
+        // let incr = get_atom(&connection, "INCR")?; // not (yet?) implemented
+
+        Ok(X11Clipboard { connection, window, atoms: Atoms { clipboard, targets, utf8_string, property } })
+    }
+
+    pub fn grab_clipboard(&self) {
+        xcb::set_selection_owner(&self.connection, self.window, self.atoms.clipboard, xcb::CURRENT_TIME);
+        self.connection.flush();
+    }
+
+    pub fn read_clipboard(&self) {
+        xcb::convert_selection(&self.connection, self.window, self.atoms.clipboard, self.atoms.utf8_string,
+                               self.atoms.property, xcb::CURRENT_TIME);
+        self.connection.flush();
+    }
+
+    pub fn run<'a>(&'a self, handle: &Handle) -> ClipboardContext<'a> {
+        let (lost_send, lost_recv) = mpsc::unbounded();
+        let (query_send, query_recv) = mpsc::unbounded();
+        let (data_send, data_recv) = mpsc::unbounded();
+        let (resp_send, resp_recv) = mpsc::unbounded();
```

Comment:
Why don't you just pass in the Controller into this function instead of having all those channels?